keps/sig-node/5233-node-readiness-gates/README.md

```diff
@@ -0,0 +1,647 @@
+    - When probes are managed by NPD: Kubelet restart does not affect the probing mechanism and condition updates will monitor the critical components.
+
+1. Network partition:
+    - The agents on the node are running and probes are succeeding locally.
+    - When there is loss of network, and the node-isolation lasts longer than the heart-beat timeout, the node-controller taints the node as ‘unhealthy’.
+    - The scheduler will therefore not schedule new pods onto this node, regardless of the status (stale or not) of the custom gate conditions. The main Ready condition acts as a fundamental gate.
+    - When network-connectivity is restored, Kubelet will update the main Ready condition to “True”. The probes will update their respective conditions. The scheduler will resume placing pods at the node once all the conditions are satisfied.
+
+#### Component Readiness Transition:
+
+1. **Component Fails:** A running component reports failure. The probing mechanism updates the corresponding node condition to `status:"False"`. The controller observes this change during its reconciliation loop. Because a required condition is no longer True, the controller re-applies the `node.kubernetes.io/not-ready:NoSchedule` taint.
+
+1. **Component Fails:** If there's a critical component rollout, and the readiness-probe detected it as health-check failure (based on thresholds and timeout configurations), it will update a blocking condition for node-scheduling.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- API validation for the new `NodeSpec.readinessGates` field.
+- Node Controller logic for taint-management based on `status.Conditions`.
+- Verify readinessGates evaluation logic.
+
+##### Integration tests
+
+* Test node registration with gates defined.
+* Test pod scheduling being blocked when gates are not met.
+* Test pod scheduling succeeding when gates transition to `True`.
+* Test existing scenarios with standard taints eg., `not-ready:NoExecute` for node timeout.
+
+##### e2e tests
+
+-  : 
+
+### Graduation Criteria
+
+* **Alpha:**
+    * Feature implemented behind `NodeReadinessGates` feature gate, default `false`.
+    * Basic unit and integration tests implemented.
+    * Initial API definition (`NodeSpec.readinessGates`) available.
+
+* **Beta:**
+    * Feature gate default `true`.
+    * e2e tests implemented and passing consistently.
+    * Scalability and performance testing performed.
```

Comment:
performance will mean some latency targets I would assume. Maybe say explicitly that we will check the latency